Source/Scene/ModelExperimental/GeometryPipelineStage.js

```diff
@@ -43,145 +92,224 @@ GeometryPipelineStage.process = function (renderResources, primitive) {
-      case VertexAttributeSemantic.TANGENT:
-        shaderBuilder.addDefine("HAS_TANGENTS");
-        break;
-      case VertexAttributeSemantic.FEATURE_ID:
-      case VertexAttributeSemantic.TEXCOORD:
-      case VertexAttributeSemantic.COLOR:
-        shaderBuilder.addDefine("HAS_" + semantic + "_" + setIndex);
-        setIndexedAttributeInitializationLines.push(
-          "    #ifdef HAS_" + semantic + "_" + setIndex
-        );
-        setIndexedAttributeInitializationLines.push(
-          "    " + varyingName + " = a_" + variableName + ";"
-        );
-        setIndexedAttributeInitializationLines.push("    #endif");
-    }
+function addAttributeToAttributesArray(
+  attributesArray,
+  attribute,
+  attributeIndex
+) {
+  var quantization = attribute.quantization;
+  var type;
+  var componentDatatype;
+  if (defined(quantization)) {
+    type = quantization.type;
+    componentDatatype = quantization.componentDatatype;
+  } else {
+    type = attribute.type;
+    componentDatatype = attribute.componentDatatype;
   }
 
   var vertexAttribute = {
     index: attributeIndex,
     value: defined(attribute.buffer) ? undefined : attribute.constant,
     vertexBuffer: attribute.buffer,
-    componentsPerAttribute: AttributeType.getNumberOfComponents(attributeType),
-    componentDatatype: attribute.componentDatatype,
+    componentsPerAttribute: AttributeType.getNumberOfComponents(type),
+    componentDatatype: componentDatatype,
     offsetInBytes: attribute.byteOffset,
     strideInBytes: attribute.byteStride,
     normalize: attribute.normalized,
   };
 
-  // Handle custom vertex attributes.
-  // For example, "_TEMPERATURE" will be converted to "a_temperature".
-  if (!defined(variableName)) {
-    variableName = attribute.name;
+  attributesArray.push(vertexAttribute);
+}
 
-    // Per the glTF 2.0 spec, custom vertex attributes must be prepended with an underscore.
-    variableName = variableName.substring(1);
-    variableName = variableName.toLowerCase();
-    varyingName = "v_" + variableName;
+function addVaryingDeclaration(shaderBuilder, attributeInfo) {
+  var variableName = attributeInfo.variableName;
+  var varyingName = "v_" + variableName;
 
-    var initializationLine =
-      "    " + varyingName + " = a_" + variableName + ";";
-    customAttributeInitializationLines.push(initializationLine);
+  var glslType;
+  if (variableName === "tangent") {
+    // Tangent's glslType is vec4, but in the shader it is split into
+    // vec3 tangent and vec3 bitangent
+    glslType = "vec3";
+  } else {
+    glslType = attributeInfo.glslType;
   }
 
-  variableName = "a_" + variableName;
   shaderBuilder.addVarying(glslType, varyingName);
+}
+
+function addAttributeDeclaration(shaderBuilder, attributeInfo) {
+  var semantic = attributeInfo.attribute.semantic;
+  var variableName = attributeInfo.variableName;
+
+  var attributeName;
+  var glslType;
+  if (attributeInfo.isQuantized) {
+    attributeName = "a_encoded_" + variableName;
```

Comment:
Would `quantized` be a better prefix for it, instead of `encoded`? It's a little tricky because with `KHR_mesh_quantization`, the `a_position` is technically quantized.

Reply:
I'll rename it. it really just pushes the naming convention problem the other way (e.g. oct encoded normals are encoded, not quantized) but the stage is named `Dequantization` stage so... I vote `quantized`